test/test_model_lazy_decoder.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2018-present MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import copy
+import unittest
+
+from itertools import chain
+
+from bson.decimal128 import Decimal128
+
+from pymodm.base.models import _LazyDecoder
+
+
+PYTHON_DATA = {
+    "python": "helloworld"
+}
+
+
+MONGO_DATA = {
+    "mongo": Decimal128("1.23")
+}
+
+
+class TestLazyDecoder(unittest.TestCase):
+    def setUp(self):
+        self.ld = _LazyDecoder()
+        self.populate_lazy_decoder(self.ld)
+
+    def tearDown(self):
+        del self.ld
+
+    def populate_lazy_decoder(self, ld):
+        for key in PYTHON_DATA:
+            ld.set_python_value(key, PYTHON_DATA[key])
+        for key in MONGO_DATA:
+            ld.set_mongo_value(key, MONGO_DATA[key])
+
+    def test_clear(self):
+        self.ld.clear()
+
+        self.assertEqual(self.ld, _LazyDecoder())
+        self.assertEqual(sum(1 for _ in self.ld), 0)
+
+    def test_iter(self):
+        all_members = set(iter(self.ld))
+        expected_members = set(
+            iter(chain(self.ld._mongo_data, self.ld._python_data))
+        )
+
+        self.assertGreater(len(expected_members), 0)
+        self.assertEqual(all_members, expected_members)
+        for item in self.ld:
+            self.assertIn(item, expected_members)
+
+    def test_eq(self):
+        ld2 = _LazyDecoder()
+        self.populate_lazy_decoder(ld2)
+        self.assertEqual(self.ld, ld2)
+
+        self.ld.set_python_value('python2', 'hellonewworld')
+        self.assertNotEqual(self.ld, ld2)
+
+    def test_pop(self):
+        def _generate_keyset(*iterables):
+            return set([k for k in iter(chain(*iterables))])
+
+        all_keys = _generate_keyset(MONGO_DATA, PYTHON_DATA)
+        expected_keyset = copy.copy(all_keys)
+        for key in all_keys:
+            self.ld.pop(key)
+            expected_keyset.discard(key)
+            data_keyset = _generate_keyset(
+                self.ld._python_data, self.ld._mongo_data)
+            iter_keyset = _generate_keyset(self.ld)
+
+            # For each check the data, iter, and contains.
+            self.assertEqual(data_keyset, expected_keyset)
+            self.assertEqual(iter_keyset, expected_keyset)
+            for candidate_key in expected_keyset:
+                self.assertTrue(candidate_key in self.ld)
+
+    def test_get_mongo_data_as_mongo_value(self):
+        ldcopy = copy.deepcopy(self.ld)
```

Comment:
`self.setUp` gets run before every test method so I don't think you should need to make a copy of it.

Reply:
The copy is made in order to preserve the starting-state of the `_LazyDecoder` object. During the test I modify `self.ld` and then I use `ldcopy` at the end to perform the necessary checks. In this case, I am simply checking for equality, i.e. the `_LazyDecoder` instance should not have changed in any way during the test.

Reply:
Oh I see.